cdap-docs/build.sh

```
@@ -20,283 +20,210 @@
-    cli_docs="${WITH}"
-  fi
-  if [ "${doc_type}" == "${DOCS_LOCAL}" ]; then
-    DOCS_LOCAL="true"
-    export DOCS_LOCAL
-  fi
+function display_end_title_bell() {
+  echo "--------------------------------------------------------"
+  ring_bell "Completed \"$1\""
   echo "========================================================"
   echo "========================================================"
-  echo "Building \"${doc_type}\""
-  echo "(${cli_docs} CLI docs, ${javadocs} Javadocs)"
-  echo "--------------------------------------------------------"
   echo
-  if [[ "${doc_type}" != "${DOCS}" && "${doc_type}" != "${DOCS_OUTER}" ]]; then
-    clean_targets
-  fi
-  clear_messages_set_messages_file
-  if [ "${doc_type}" != "${DOCS_OUTER}" ]; then
-    run_command docs-first-pass ${source_path}
-  fi
-  if [ "${doc_type}" == "${DOCS}" -o "${doc_type}" == "${DOCS_OUTER}" ]; then
-    if [ "${javadocs}" != "${WITH}" -o "${cli_docs}" != "${WITH}" ]; then
-      check_build_rst
-    fi
-    build_docs_outer_level ${source_path}
-    copy_docs_inner_level
-  else
-    clear_messages_set_messages_file
-  fi
-  if [ "${javadocs}" == "${WITH}" -o "${cli_docs}" == "${WITH}" ]; then
-    if [ "${javadocs}" == "${WITH}" ]; then
-      run_command javadocs
-    fi
-    if [ "${cli_docs}" == "${WITH}" ]; then
-      run_command docs-cli
-    fi
-  else
-    check_build_rst
-  fi
-  if [ "${doc_type}" == "${GITHUB}" -o "${doc_type}" == "${GITHUB_ONLY}" ]; then
-    run_command docs-github-part ${source_path}
-  elif [ "${doc_type}" == "${WEB}" -o "${doc_type}" == "${WEB_ONLY}" ]; then
-    run_command docs-web-part ${source_path}
-  fi
-  set_display_version
-  display_messages_file
+}
+
+function build_docs_set() {
+  local warnings
+  local title="Building Docs Set"
+  display_start_title ${title}
+  
+  clean_targets
+  clear_messages_file
+  build_docs_first_pass
+  clear_messages_file
+  build_javadocs docs
+  build_docs_cli
+  build_docs_second_pass
+  build_docs_package
```

What if any of these fail?

---

I don't consider it a huge problem; I would rather run through them all, and get all the errors in a single pass, rather than having to keep running over and over, fixing the new erros that surface. Typically one section will have errors, but other parts will pass.

---

OK, but this code isn't written in a way for that to make sense. A failure in one of these functions will cause cascading failures/problems. They're not written as fully independent units of work. Therefore, the script should fail fast, rather than continue generating false-positive errors (along with actual errors, yes) and using up build resources.

---

I'm not sure about that: any of those four can fail without it cascading to the next. The Javadocs and the CLI docs can be (and are) ignored if they fail or aren't built, and if there is a failure in the second_pass, you can still package what was copied. 

I looked over these functions, and I'm not sure what errors I would promulgate, or what codes I would pass that would be meaningful. Many of the functions in these functions don't return error codes themselves, as they don't really have an impact outside themeselves. We have a system that captures Sphinx and build errors and puts them in a file that gets read and reported on at the end, and I think that is working well enough. 

I'm having trouble seeing the benefit to spending a lot of time here.

---

I'm reading that it's acceptable to package incomplete/partial builds, versus failing. If that's the case, I'm going to just let this go. It doesn't make any sense to me, and you haven't been able to explain why failure is acceptable. To me, the end result is bogus. If files that are supposed to be copied, aren't, then how is downstream supposed to know that something's gone wrong? If this were in some kind of test suite, that would be one thing. However, this is a build. Single failures should be enough to be fatal and it should fail fast. This is a complete anti-pattern and not something to be encouraged, IMO.

At any rate, it's not worth fighting over. Feel free to merge without fixing this.

---

I see your point. Maybe we need it both ways: when this is run as build+deploy (or as a build on Bamboo) it needs to behave as you describe, but locally (on a Mac) allow that behaviour be over-ridden.

---

Not for this PR, but later, you could implement a "test_or_die" function or something, which checks if tests should be enabled, and dies if not. This would allow both sets of functionality. 😄

---

As we need to get this done, I will merge with this behavour, but add an issue to correct this: https://issues.cask.co/browse/CDAP-7573